### PR TITLE
[Security] Fix remote DoS via invalid recovered token reinjection

### PR DESCRIPTION
**Pull request by jperezdealgaba**
## Summary

Fixes https://github.com/vllm-project/vllm/security/advisories/GHSA-8wr5-jm2h-8r4f

When `vocab_size` is not a multiple of `BLOCK_SIZE` (8192), the last Triton tile extends beyond the vocabulary. If all valid entries in that tail tile have zero target probability, the out-of-range masked positions (score 0) could win the `tl.max` tie-break, producing `recovered_id >= vocab_size`. This results in a remote denial of service via invalid token reinjection.

**Fix:**
- Mask out-of-vocabulary entries to `-inf` in `sample_recovered_tokens_kernel` so they can never win the tiled argmax
- Add a belt-and-suspenders clamp on `recovered_id` before store

## Test plan

- Added regression test `test_sample_recovered_tokens_vocab_boundary` in `tests/v1/sample/test_rejection_sampler.py` that forces the zero-tail-tile scenario for non-aligned vocab sizes (8193, 10000, 151936)
- Verified every recovered token ID is within `[0, vocab_size)`

## Notes

- AI assistance was used in developing this fix (per AGENTS.md)
- This fix was developed in the GHSA private fork and is now submitted for public review

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Review by MatthewBonanni — APPROVED** [review]
Thanks for the fix, this seems reasonable to me. Can you confirm that the test fails on the unpatched kernel though? Shouldn't the failure happen when `vocab_size < BLOCK_SIZE`?

**Comment by jperezdealgaba** [comment]
@MatthewBonanni Yes, the test fails on the unpatched kernel -- without the -inf masking, the zero-score out-of-range entries in the last partial tile can win the tl.max tie-break, producing recovered_id >= vocab_size.

The bug applies when vocab_size < BLOCK_SIZE too (single partial tile, same mechanism). The fix handles both cases since vocab_mask = vocab_offset < vocab_size works regardless of whether vocab_size is above or below BLOCK_SIZE.

I have added also a test case for that specific use case

File: vllm/v1/sample/rejection_sampler.py [hunk 1]
```diff
@@ -921,12 +921,17 @@ def sample_recovered_tokens_kernel(
...
+        # when all valid entries in the last tile have zero probability.
         score = prob * inv_q
+        score = tl.where(vocab_mask, score, float("-inf"))
         local_max, local_id = tl.max(score, axis=0, return_indices=True)
 
         if local_max > max_val:
             max_val = local_max
             recovered_id = v + local_id
 
+    recovered_id = tl.minimum(recovered_id, vocab_size - 1)
```
Comment (ekagra-ranjan) [inline vllm/v1/sample/rejection_sampler.py:936]: wondering what's the original motivation for this when the code earlier maks the OOV tokens with -inf so they are not sampled?

On the contrary, I think at this line we should make any OOV token as 0 as per this year old [PR](https://github.com/vllm-project/vllm/pull/21876). That PR found that nans in the logits will make the target sample tokenid 0 whereas triton will sample OOV since argmax of [nan, nan, -inf, -inf] is 2. The current vLLM code can still lead to OOV from draft until we add this fix. The year old PR doesn't fix the vLLM's behavior of outputting tokenid 0 when nans are in logits but rather makes draft behave like target. As of now, nans in logits gets unnoticed in non SD inference and without this fix it gives an impression that SD is breaking something. 

As to why nans can be present in model, I can think of 2 scenarios: some quant method OR some early stage RL checkpoints.

**Review by ekagra-ranjan — COMMENTED** [review]
Thanks for the PR @jperezdealgaba !

I have a followup on a pending issue which is similar to what was observed a year ago but hasn't been fixed on vLLM and is related to the fix in this PR so curious to know your thoughts on below @MatthewBonanni @jperezdealgaba
